Add delete tag endpoint (#8)
mlflow/mlflow-go: docs/porting-a-new-endpoint.md

# Porting a New Endpoint

In order to complete the Go implementation of the tracking server, we need to port the missing endpoints. 
The following guide will walk you through the whole process of viewing the missing endpoints, implementing them and making sure they work with MLflow's existing integration tests.

1. [View Missing Endpoints](#view-missing-endpoints): View the current state of implementated endpoints.
2. [Enable the Endpoint](#enable-the-endpoint): Enable code generation for a new endpoint.
3. [Validate Input](#validate-input): Apply validation on the incoming request data.

> 4. [Implement the Service Method](#implement-the-service-method): Adding a new method to the TrackingService interface allows handling the logic for the new endpoint

Reviewer: ```suggestion
4. [Implement the Service Method](#implement-the-service-method): Add a new method to the `TrackingService` interface to handle the logic for the new endpoint
```

	service, err := trackingServices.Get(serviceID)
	if err != nil {
		return makePointerFromError(err, responseSize)
	}
	return invokeServiceMethod(service.DeleteTag, new(protos.DeleteTag), requestData, requestSize, responseSize)
}
```

We will later create a Python binding to enable us to invoke this generated Go code.

## Validate Input

A first step to port an endpoint would be to check the request validation happening at the HTTP level. Open [handlers.py](../.mlflow.repo/mlflow/server/handlers.py) (from the `.mlflow.repo`) to see which fields are required.

```python
@catch_mlflow_exception
@_disable_if_artifacts_only
def _delete_tag():
    request_message = _get_request_message(
        DeleteTag(),
        schema={
            "run_id": [_assert_required, _assert_string],
            "key": [_assert_required, _assert_string],
        },
    )
    _get_tracking_store().delete_tag(request_message.run_id, request_message.key)
    response_message = DeleteTag.Response()
    response = Response(mimetype="application/json")
    response.set_data(message_to_json(response_message))
    return response
```

> We notice that `run_id` and `key` are required. Too avoid writing repeatitive code in Go, we make use of [Go validator](https://github.com/go-playground/validator) and use annotations to validate the incoming structs.

Reviewer: ```suggestion
We notice that `run_id` and `key` are required. To avoid writing repetitive code in Go, we make use of [Go validator](https://github.com/go-playground/validator) and use annotations to validate the incoming structs.
```

To ensure parity, we need to update the input struct `DeleteTag` from the generated proto code ([pkg/protos/service.pb.go](../pkg/protos/service.pb.go)) with annotations. ⚠️ Since this file was generated, we don't want to modify it directly. ⚠️ Instead, we will configure validations in [validations.go](../magefiles/generate/validations.go).

```go
var validations = map[string]string{
    ...
	"DeleteTag_RunId":                    "required",
	"DeleteTag_Key":                      "required",
}
```

After that, run `mage generate` again and check if our fields in `DeleteTag` now contain `validate:"required"`.

> Note that we do not need to process `_assert_string` because in Go all the structs are already statically typed.

Reviewer: ```suggestion
Note that we do not need to process `_assert_string` because in Go, all structs are already statically typed.
```

It may happen that the [baked-in validations](https://github.com/go-playground/validator?tab=readme-ov-file#baked-in-validations) of the Go validator are not sufficient. If this occurs, please create a [custom validation rule](https://pkg.go.dev/github.com/go-playground/validator/v10#hdr-Custom_Validation_Functions) in [pkg/validation/validation.go](../pkg/validation/validation.go).

> ## Implement the Service Method

Reviewer: ```suggestion
## Implement the Service Logic
```

We aim to keep the Go implementation as close to the Python code as possible. That's why we have a thin service layer and keep most logic in the (SQL) store instead of the service. This keeps things similar to Python, making it easier to port and compare the code.

In the [pkg/tracking/service/service.go](../pkg/tracking/service/service.go), we need to call the store with the same arguments as the Python store would be called:

```python
# handlers.py
_get_tracking_store().delete_tag(request_message.run_id, request_message.key)
```

> We split up the service over multiple files (one Go file per entity), sometimes it might not existing yet.

Reviewer: ```suggestion
We split up the service over multiple files (one Go file per entity), sometimes it might not exist yet.
```

### Go unit tests

An example use case where unit tests proved to be highly beneficial is the `filter` and `order` parser in the [search-runs](https://mlflow.org/docs/latest/rest-api.html#search-runs) endpoint. The [query_test.go](../pkg/tracking/service/query/query_test.go) file tests the more complex implementation of the endpoint, which has no external dependencies.

## Run Tests

Run `mage test:python` and verify that our Go implementation passes the existing tests.

There is one caveat to these tests; occasionally, they may have a Python bias, meaning that the tests pass due to Python's dynamic nature, while our Go tests might fail because they are strongly typed. Another issue may arise if the Python implementation does not consistently return the same error messages. Therefore, it may be necessary to submit a PR to [mlflow](https://github.com/mlflow/mlflow) to adjust the existing tests.

Some examples include:
- https://github.com/mlflow/mlflow/pull/13233
- https://github.com/mlflow/mlflow/pull/13128
- https://github.com/mlflow/mlflow/issues/12550

## Troubleshooting

> If you encounter any difficulties, please feel free to open a draft PR and ask specific questions. Once your questions are answered, kindly update this section if you’ve learned something that should be included in this guide.

Reviewer: ```suggestion
If you encounter any difficulties, please feel free to open a draft PR and ask specific questions. Once your questions are answered, kindly update this section if you’ve learned something that could be valuable for other contributors.
```
